### ISSUE TITLE
Pluggable Resolvers Architecture

### ISSUE THREAD

**Issue by jbaruch** [bower/bower#1649]
# Motivation
Bower resolvers mechanism is part of the Resolve process and is coupled with the Bower core; Therefore, extending Bower with custom (external) resolvers requires extending the Bower core project and having any resolver maintained by the Bower team.
In order to overcome this limitation, Bower needs a plugin-in system for resolvers, that will make it more flexible to extend Bower with custom resolvers and will allow external contributors to maintain these plugins.

# Proposed Implementation Details
`.bowerrc`
A new section will be added to the configuration file, that will contain npm <package_name>:<version> of external resolvers, added by users. When no version is specified, the latest version will be used.
For example:

```json
       "resolvers":[
           "custom-resolver",
           "x-resolver:1.0.0",
       ]
```

*Open discussion:*
 1. Considering the external resolvers order in the list

# External Resolvers
 * A resolver will be independent NPM package, maintained and under the responsibility of a contributor.
 * As before, will implements the `Resolver` ("abstract class") methods, but with an addition of a new standard method that will decide if the current resolver `matches` the `source` (URL, git endpoint, etc.). For example: `_matches(source)`

*Open discussion:*
 1. Should existing resolver logic and utils be extracted to external packages? This will allow using dependencies by contribute resolvers cleanly, without relative `require()` to Bower files location (`require('../../../lib/core/resolvers/Resolver')`) 

# Resolver Factory   
Main flow:
 1. Will go over the external resolvers in the `.bowerrc`, and will try to load them as Node modules (`require`)
 2. For every module, invoke the resolver `_match` method, that will decide if the current resolver matches the `source` type
 3. In case a resolver matches, use it (return it to `PackageRepository`)
 4. In case no external resolver matches the source (or there aren't any), fall back to the list of internal resolvers (`UrlResolver`, `GitRemoteResolver`, etc.)

*Open discussion:*
 1. Should `ResolverFactory` npm automatically install external resolvers in **runtime**, in case it cannot find a resolver installed in Bower. Npm can install the resolver to the `node_modules` folder under Bower package

  **Cons:**
   * `npm install` behind the scenes without the user approval. Also counting on the default npm configuration. Should we support custom npm configuration?
   * Should auto-update of installed resolvers be allowed? Maybe a flag in `.bowerrc`

  **Pros:**
   * Applying external resolvers with zero setup and zero config

**Comment by jbaruch** [comment]
Summoning @yoav @drorb @liorhson

**Comment by sheerun** [comment]
Hey!

1. In my opinion putting resolver in dependencies of `package.json` should be enough. Bower automatically requires resolver if necessary. This means no entry in `.bowerc` is requred.
2. Instead resolver matching use resolver prefix for dependency, just like JSPM does. e.g. `bitbucket:jquery/jquery` (it requires resolver with `require('bower-bitbucket-resolver')`, or maybe finds it some other way, e.g. based on meta info).
3. Yes, I think we should extract existing resolvers.
4. No, I don't think resolver should be installed automatically. Public packages should depend only on bower-provided default resolvers. You can put custom one explicitly in `package.json`.

What about resolver interface?

**Comment by liorhson** [comment]
Sounds good,

1. Are you referring to bower's **own** `package.json`? That implies that user will need to update the list of dependencies with each new bower version. Is that ok?
2. So, if I understand correctly, every resolver would have a prefix that is used when declaring a bower dependency. How would you suggest to translate this npm package name to `require()`? Some form of naming convention, like `bower-<resolver-prefix>-resolver`?
3. We need to maintain some backward compatibility with existing registry URL values, no? How do you suggest we do that? Probably the bower **registry** need to supports this prefix convention also, for example: `git:<URL>`, `svn:<URL>`, etc.

The resolver interface would stay pretty the same, wouldn't it? I can submit a proposal for the fine details, if you wish.

Jfrog

**Comment by sheerun** [comment]
1. Project's `package.json`. Bower's `package.json` can contain default resolvers.
2. I'm not sure about this. Either naming convention or metadata in `package.json` of resolver (like `"bower_resolver"` tag). I think it's up to discussion. I wonder how other packages handle plugins.
3. Of course we need to :) Practically all of registry's prefixes are either `git:` or `svn:`. Public registry won't accept custom prefixes (only from default resolvers), but you can create custom registry.

I think resolver's interface shouldn't stay the same. It's pretty messy, and coupled with bower. We have good opportunity to redesign API for resolvers and make them bower-independent. Good first step is to determine core & optional features for resolvers. I'd also love interface to be based on Promises.

Adam

**Comment by liorhson** [comment]
Following our conversation, I want to summary some of the things we talked about:

## Extraction
Extract needed features from the Bower core, in order to use them in the pluggable resolvers. The   main features are the Bower **utils**, **renderers**, and the extendable Base `Resolver`.

For POC, I created  repositories for the Utils, and the Base Resolver as standalone npm packages. 
   - https://github.com/liorhson/bower-utils
   - https://github.com/liorhson/bower-resolver-base

Also POC Bower project repository, without the above features in the core (the tests are not modified yet), and add them as a dependencies in the Bower package.json.
```json
"dependencies": {
    "bower-utils":"*",
    "bower-resolver-base":"*"
}
```
   - https://github.com/liorhson/bower

Pluggable/custom resolvers, will be a independent npm packages, that will contains the bower project in the **peerDependencies**. The `resolverFactory` will recognize them with the registry prefix source for example: `**bitbucket**:jquery/jquery`.
Of course we will need backward compatibility with old registry entries, that doesn`t support the new mechanism.
Something like:
```javascript
    var pkgResolver;
    var registryPrefix = extractResolverPrefix(source);

    if(registryPrefix) {
        try {
            pkgResolver = require('bower-' + registryPrefix + '-resolver');

            return Q.fcall(function () {
                return [pkgResolver, source];
            });
        }
        catch (err) {
            //go to backward compatibility resolvers...
        }
    }

    //go to backward compatibility resolvers...

```

Default repositories (`GitHubResolver`, `SvnResolver`, etc.) will be also extract from the Bower core, and will be exists as dependencies in the Bower package.json.
```json
"dependencies": {
    "git-hub-resolver":"*",
    "svn-resolver":"*"
}
```

## Interface
The resolver interface will change for some improvements:

`resolver.download(url, config = {})` something like 
```
Resolver.download(
  'github:jquery/jquery', '1.2.3', { foo: 'bar' }
).then(function(tmpDir) {
  console.log(tmpDir);
});
```
Possible return types: 
 * `result: Promise & EventEmitter`
 * `{ logger: EventEmitter } : Promise`
 * `{ promise: Promise }: EventLogger`

`resolver.versions(endpoint, config = {})`, that returns promises of available versions of the given endpoint, and also the calculation of the caching key will be decided by the resolver itself.
Return type, something like `[{ version: '*', cacheKey: 'etag12345' }]`

Overall view:
![bower-arch](https://cloud.githubusercontent.com/assets/6830400/5755333/d56662c4-9cab-11e4-976c-c44924ae8ce4.jpg)

**Comment by sheerun** [comment]
Implemented in #1871
